src/ai/defaultai.cc

```diff
@@ -850,31 +857,76 @@ void DefaultAI::late_initialization() {
 				}
 			}
 
-			// some important buildings are identified first the woodcutter/lumberjack
-			if (bh.collects_ware_from_map() == "log") {
-				bo.set_is(BuildingAttribute::kLumberjack);
-			}
-			// quarries
-			if (bh.collects_ware_from_map() == "granite") {
-				bo.set_is(BuildingAttribute::kNeedsRocks);
-			}
-			// wells
-			if (bh.collects_ware_from_map() == "water") {
-				bo.set_is(BuildingAttribute::kWell);
+			// Some important buildings are identified
+			if (prod.input_wares().empty() && !prod.output_ware_types().empty() &&
+			    prod.created_immovables().empty() && !prod.collected_immovables().empty()) {
+				bool produces_construction_material = false;
+				bool produces_non_construction_material = false;
+				for (Widelands::DescriptionIndex output_idx : prod.output_ware_types()) {
+					if (tribe_->is_construction_material(output_idx)) {
+						produces_construction_material = true;
+					} else {
+						produces_non_construction_material = true;
+					}
+				}
+				if (produces_construction_material) {
+					// TODO(GunChleoc): We should lose the hard distinction between quarry and
+					// lumberjack, so that a building can be both
+					if (prod.supported_by_productionsites().empty()) {
```

Comment:
Why dont we unify this logic below, I mean match attribute from collected_attributes directly to (strings?) rocks, logs and berries. Right now here is rather a indirect guessing than direct identification of what collected_attribute it is.

Also prod.collected_attributes can return more than one item, correct? Can one building collect more than one type of objects from map?

Reply:
for the first question: this was designed by @GunChleoc so I cn only tell what might have been the reason. I guess it was because this would result in some hardcoding again while this attempt should work without hardcoding the relations you mentioned. 
Second question: yes there are some buildings: amazons huntergatherer collects a ressource (fish) and bobs (game).  Frisians charcoal burner hut "collects" empty ponds and ponds with charcoal.

Reply:
I must say I am quite lost in this logic. This is also because I do not know new tribes.

Reply:
you can have a look in GunChleocs original branch where there was some debug code to print the buildings identified. I removed this for review however this could give you a good impression of this logic